File: bitePal_ui/lib/models/recipe.dart
/// 菜谱模型
class Recipe {
  /// 菜谱ID
  final String id;

  /// 菜谱名称
  final String name;

  /// 图片URL
  final String? image;

  /// 制作时间
  final String time;

  /// 难度（简单/中等/困难）
  final String difficulty;

  /// 难度颜色（从数据库读取）
  final String? difficultyColor;

  /// 标签数组
  final List<String> tags;

  /// 标签颜色数组
  final List<String> tagColors;

  /// 是否收藏
  final bool favorite;

  /// 分类数组
  final List<String> categories;

  /// 食材列表
  final List<Ingredient>? ingredients;

  /// 制作步骤
  final List<String>? steps;

  /// 创建用户ID
  final String? userId;

  /// 是否公开
  final bool? isPublic;

  Recipe({
    required this.id,
    required this.name,
    this.image,
    required this.time,
    required this.difficulty,
    this.difficultyColor,
    required this.tags,
    required this.tagColors,
    this.favorite = false,
    required this.categories,
    this.ingredients,
    this.steps,
    this.userId,
    this.isPublic,
  });

  /// 从JSON创建Recipe实例
  /// json: JSON数据
  /// 返回: Recipe实例
  factory Recipe.fromJson(Map<String, dynamic> json) {
    return Recipe(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? '',
      image: json['image'],
      time: json['time'] ?? '',
      difficulty: json['difficulty'] ?? '简单',
      difficultyColor: json['difficultyColor'],
      tags: List<String>.from(json['tags'] ?? []),
      tagColors: List<String>.from(json['tagColors'] ?? []),
      favorite: json['favorite'] ?? false,
      categories: List<String>.from(json['categories'] ?? []),
      ingredients: json['ingredients'] != null
          ? (json['ingredients'] as List)
                .map((e) => Ingredient.fromJson(e))
                .toList()
          : null,
      steps: json['steps'] != null ? List<String>.from(json['steps']) : null,
      userId: json['userId'],
      isPublic: json['isPublic'],
    );
  }

  /// 转换为JSON
  /// 返回: JSON Map
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'image': image,
      'time': time,
      'difficulty': difficulty,
      'difficultyColor': difficultyColor,
      'tags': tags,
      'tagColors': tagColors,
      'favorite': favorite,
      'categories': categories,
      'ingredients': ingredients?.map((e) => e.toJson()).toList(),
      'steps': steps,
      'isPublic': isPublic,
    };
  }

  /// 复制并修改
  Recipe copyWith({
    String? id,
    String? name,
    String? image,
    String? time,
    String? difficulty,
    String? difficultyColor,
    List<String>? tags,
    List<String>? tagColors,
    bool? favorite,
    List<String>? categories,
    List<Ingredient>? ingredients,
    List<String>? steps,
    String? userId,
    bool? isPublic,
  }) {
    return Recipe(
      id: id ?? this.id,
      name: name ?? this.name,
      image: image ?? this.image,
      time: time ?? this.time,
      difficulty: difficulty ?? this.difficulty,
      difficultyColor: difficultyColor ?? this.difficultyColor,
      tags: tags ?? this.tags,
      tagColors: tagColors ?? this.tagColors,
      favorite: favorite ?? this.favorite,
      categories: categories ?? this.categories,
      ingredients: ingredients ?? this.ingredients,
      steps: steps ?? this.steps,
      userId: userId ?? this.userId,
      isPublic: isPublic ?? this.isPublic,
    );
  }
}

/// 食材模型
class Ingredient {
  /// 食材名称
  final String name;

  /// 用量
  final String amount;

  /// 是否可用
  final bool available;

  Ingredient({required this.name, required this.amount, this.available = true});

  /// 从JSON创建Ingredient实例
  factory Ingredient.fromJson(Map<String, dynamic> json) {
    return Ingredient(
      name: json['name'] ?? '',
      amount: json['amount'] ?? '',
      available: json['available'] ?? true,
    );
  }

  /// 转换为JSON
  Map<String, dynamic> toJson() {
    return {'name': name, 'amount': amount, 'available': available};
  }
}

File: bitePal_service/models/recipe.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// StringArray 字符串数组类型（用于GORM存储）
type StringArray []string

// Scan 从数据库读取值
func (s *StringArray) Scan(value interface{}) error {
	if value == nil {
		*s = StringArray{}
		return nil
	}

	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		*s = StringArray{}
		return nil
	}

	return json.Unmarshal(bytes, s)
}

// Value 写入数据库的值
func (s StringArray) Value() (driver.Value, error) {
	return json.Marshal(s)
}

// RecipeIngredient 菜谱食材结构
type RecipeIngredient struct {
	Name      string `json:"name"`      // 食材名称
	Amount    string `json:"amount"`    // 用量
	Available bool   `json:"available"` // 是否可用
}

// RecipeIngredients 菜谱食材数组类型
type RecipeIngredients []RecipeIngredient

// Scan 从数据库读取值
func (r *RecipeIngredients) Scan(value interface{}) error {
	if value == nil {
		*r = RecipeIngredients{}
		return nil
	}

	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		*r = RecipeIngredients{}
		return nil
	}

	return json.Unmarshal(bytes, r)
}

// Value 写入数据库的值
func (r RecipeIngredients) Value() (driver.Value, error) {
	return json.Marshal(r)
}

// Recipe 菜谱模型
type Recipe struct {
	ID              string            `json:"id" gorm:"type:varchar(36);primaryKey"`               // 菜谱ID
	Name            string            `json:"name" gorm:"type:varchar(100);not null;index"`        // 菜谱名称
	Image           string            `json:"image" gorm:"type:varchar(500)"`                      // 图片URL
	Time            string            `json:"time" gorm:"type:varchar(50)"`                        // 制作时间
	Difficulty      string            `json:"difficulty" gorm:"type:varchar(50)"`                  // 难度（有手就行/家常便饭/餐厅招牌/硬核挑战/专业厨师）
	DifficultyColor string            `json:"difficultyColor" gorm:"-"`                            // 难度颜色（不存储，从分类表查询）
	Tags            StringArray       `json:"tags" gorm:"type:jsonb"`                              // 标签数组
	TagColors       StringArray       `json:"tagColors" gorm:"type:jsonb"`                         // 标签颜色数组
	Favorite        bool              `json:"favorite" gorm:"default:false"`                       // 是否收藏
	Categories      StringArray       `json:"categories" gorm:"type:jsonb"`                        // 分类数组
	Ingredients     RecipeIngredients `json:"ingredients" gorm:"type:jsonb"`                       // 食材列表
	Steps           StringArray       `json:"steps" gorm:"type:jsonb"`                             // 制作步骤
	UserID          string            `json:"userId" gorm:"type:varchar(36);column:user_id;index"` // 创建用户ID
	IsPublic        bool              `json:"isPublic" gorm:"default:false"`                       // 是否公开
	CreatedAt       time.Time         `json:"createdAt"`                                           // 创建时间
	UpdatedAt       time.Time         `json:"updatedAt"`                                           // 更新时间
	DeletedAt       gorm.DeletedAt    `json:"-" gorm:"index"`                                      // 软删除时间
}

// BeforeCreate 创建前钩子，自动生成ID
func (r *Recipe) BeforeCreate(tx *gorm.DB) error {
	if r.ID == "" {
		r.ID = uuid.New().String()
	}
	return nil
}

// RecipeListItem 菜谱列表项（简化版本）
type RecipeListItem struct {
	ID              string      `json:"id"`              // 菜谱ID
	Name            string      `json:"name"`            // 菜谱名称
	Image           string      `json:"image"`           // 图片URL
	Time            string      `json:"time"`            // 制作时间
	Difficulty      string      `json:"difficulty"`      // 难度
	DifficultyColor string      `json:"difficultyColor"` // 难度颜色
	Tags            StringArray `json:"tags"`            // 标签数组
	TagColors       StringArray `json:"tagColors"`       // 标签颜色数组
	Favorite        bool        `json:"favorite"`        // 是否收藏
	Categories      StringArray `json:"categories"`      // 分类数组
}

// ToListItem 转换为列表项
// 返回: 菜谱列表项
func (r *Recipe) ToListItem() *RecipeListItem {
	return &RecipeListItem{
		ID:              r.ID,
		Name:            r.Name,
		Image:           r.Image,
		Time:            r.Time,
		Difficulty:      r.Difficulty,
		DifficultyColor: r.DifficultyColor,
		Tags:            r.Tags,
		TagColors:       r.TagColors,
		Favorite:        r.Favorite,
		Categories:      r.Categories,
	}
}

// RecipeType 菜谱类型常量
const (
	RecipeTypeMyRecipe     = "my_recipe"     // 我的私房菜
	RecipeTypePublicRecipe = "public_recipe" // 公开菜谱
)

// UserFavorite 用户收藏关联表
type UserFavorite struct {
	ID         string         `json:"id" gorm:"type:varchar(36);primaryKey"`                                        // ID
	UserID     string         `json:"userId" gorm:"type:varchar(36);not null;uniqueIndex:idx_user_recipe_type"`     // 用户ID
	RecipeID   string         `json:"recipeId" gorm:"type:varchar(36);not null;uniqueIndex:idx_user_recipe_type"`   // 菜谱ID
	RecipeType string         `json:"recipeType" gorm:"type:varchar(20);not null;uniqueIndex:idx_user_recipe_type"` // 菜谱类型
	CreatedAt  time.Time      `json:"createdAt"`                                                                    // 创建时间
	DeletedAt  gorm.DeletedAt `json:"-" gorm:"index"`                                                               // 软删除时间
}

// BeforeCreate 创建前钩子，自动生成ID
func (uf *UserFavorite) BeforeCreate(tx *gorm.DB) error {
	if uf.ID == "" {
		uf.ID = uuid.New().String()
	}
	return nil
}
